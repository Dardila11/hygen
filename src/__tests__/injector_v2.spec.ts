import injectorV2 from '../ops/injector_v2'

const gemfile = `
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    `

const configFile = `
config:
  database: mysql
  cache: redis
  logging: true
`

const packageJson = `{
  "name": "my-app",
  "version": "1.0.0",
  "dependencies": {
    "react": "^18.0.0"
  }
}`

describe('injectorV2', () => {
  describe('location-based injection', () => {
    it('before rails', () => {
      expect(
        injectorV2(
          {
            attributes: {
              before: "gem 'rails'",
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('after rails', () => {
      expect(
        injectorV2(
          {
            attributes: {
              after: "gem 'rails'",
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('prepend top of file', () => {
      expect(
        injectorV2(
          {
            attributes: {
              prepend: true,
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('append bottom of file', () => {
      expect(
        injectorV2(
          {
            attributes: {
              append: true,
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('at_line 2 (below "source")', () => {
      expect(
        injectorV2(
          {
            attributes: {
              at_line: 2,
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('returns original content when no valid location attribute found', () => {
      expect(
        injectorV2(
          {
            attributes: {},
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toBe(gemfile)
    })

    it('returns original content when pattern not found', () => {
      expect(
        injectorV2(
          {
            attributes: {
              before: 'nonexistent pattern',
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toBe(gemfile)
    })
  })

  describe('skip_if functionality', () => {
    it('skip_if "source" exists', () => {
      expect(
        injectorV2(
          {
            attributes: {
              skip_if: 'source',
              after: "gem 'rails'",
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('skip_if pattern does not exist - should inject', () => {
      expect(
        injectorV2(
          {
            attributes: {
              skip_if: 'nonexistent',
              after: "gem 'rails'",
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('correctly interpret multi-line skip_if regex', () => {
      expect(
        injectorV2(
          {
            attributes: {
              skip_if: "rails[a-z\\:\\/\\.'\\s]*giri",
              after: "gem 'rails'",
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('skip_if with multiline flag works correctly', () => {
      const multilineContent = `line1
line2
line3`
      expect(
        injectorV2(
          {
            attributes: {
              skip_if: '^line2$',
              prepend: true,
            },
            body: 'new line',
          },
          multilineContent,
        ),
      ).toBe(multilineContent) // Should be skipped
    })
  })

  describe('replace functionality', () => {
    it('replaces single occurrence', () => {
      expect(
        injectorV2(
          {
            attributes: {
              replace: 'mysql',
            },
            body: 'postgresql',
          },
          configFile,
        ),
      ).toMatchSnapshot()
    })

    it('replaces multiple occurrences with global flag', () => {
      const content = `
server: mysql
backup: mysql
cache: redis
`
      expect(
        injectorV2(
          {
            attributes: {
              replace: 'mysql',
            },
            body: 'postgresql',
          },
          content,
        ),
      ).toMatchSnapshot()
    })

    it('replaces with regex pattern', () => {
      expect(
        injectorV2(
          {
            attributes: {
              replace: '"react":\\s*"[^"]*"',
            },
            body: '"react": "^19.0.0"',
          },
          packageJson,
        ),
      ).toMatchSnapshot()
    })

    it('returns original content when replace pattern not found', () => {
      expect(
        injectorV2(
          {
            attributes: {
              replace: 'nonexistent',
            },
            body: 'replacement',
          },
          configFile,
        ),
      ).toBe(configFile)
    })

    it('replace takes precedence over location attributes', () => {
      expect(
        injectorV2(
          {
            attributes: {
              replace: 'redis',
              after: 'database',
            },
            body: 'memcached',
          },
          configFile,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('eof_last functionality', () => {
    it('if eof_last is false remove empty line from the end of injection body', () => {
      expect(
        injectorV2(
          {
            attributes: {
              after: "gem 'rails'",
              eof_last: false,
            },
            body: "    gem 'kamikaze' # added by hygen\n",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('if eof_last is true insert empty line to injection body', () => {
      expect(
        injectorV2(
          {
            attributes: {
              after: "gem 'rails'",
              eof_last: true,
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('if eof_last is false with carriage return', () => {
      expect(
        injectorV2(
          {
            attributes: {
              after: "gem 'rails'",
              eof_last: false,
            },
            body: "    gem 'kamikaze' # added by hygen\r\n",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('body without newline and eof_last true gets newline added', () => {
      expect(
        injectorV2(
          {
            attributes: {
              prepend: true,
              eof_last: true,
            },
            body: '# Header comment',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('body with newline and eof_last undefined preserves newline', () => {
      expect(
        injectorV2(
          {
            attributes: {
              prepend: true,
            },
            body: '# Header comment\n',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('multi-line pattern matching', () => {
    it('correctly interpret multi-line after regex', () => {
      expect(
        injectorV2(
          {
            attributes: {
              after: "rails[a-z\\:\\/\\.'\\s]*giri",
              eof_last: false,
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('correctly interpret multi-line before regex', () => {
      expect(
        injectorV2(
          {
            attributes: {
              before: "rails[a-z\\:\\/\\.'\\s]*giri",
              eof_last: false,
            },
            body: "    gem 'kamikaze' # added by hygen",
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('handles complex multi-line patterns spanning multiple lines', () => {
      const complexContent = `
function setup() {
  const config = {
    database: 'mysql',
    host: 'localhost'
  };
  return config;
}
`
      expect(
        injectorV2(
          {
            attributes: {
              after: 'config\\s*=\\s*\\{[^}]*\\}',
            },
            body: "  console.log('Setup complete');",
          },
          complexContent,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('newline handling', () => {
    it('preserves Unix line endings', () => {
      const unixContent = 'line1\nline2\nline3\n'
      const result = injectorV2(
        {
          attributes: {
            after: 'line2',
          },
          body: 'inserted',
        },
        unixContent,
      )
      expect(result).toContain('\n')
      expect(result).not.toContain('\r\n')
    })

    it('preserves Windows line endings', () => {
      const windowsContent = 'line1\r\nline2\r\nline3\r\n'
      const result = injectorV2(
        {
          attributes: {
            after: 'line2',
          },
          body: 'inserted',
        },
        windowsContent,
      )
      expect(result).toContain('\r\n')
    })

    it('handles mixed line endings', () => {
      const mixedContent = 'line1\nline2\r\nline3\n'
      const result = injectorV2(
        {
          attributes: {
            after: 'line2',
          },
          body: 'inserted',
        },
        mixedContent,
      )
      // Should detect and use the first line ending style found
      expect(result).toContain('inserted')
    })
  })

  describe('edge cases', () => {
    it('handles empty content', () => {
      expect(
        injectorV2(
          {
            attributes: {
              prepend: true,
            },
            body: 'first line',
          },
          '',
        ),
      ).toMatchSnapshot()
    })

    it('handles empty body', () => {
      expect(
        injectorV2(
          {
            attributes: {
              after: "gem 'rails'",
            },
            body: '',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('handles content with only one line', () => {
      expect(
        injectorV2(
          {
            attributes: {
              append: true,
            },
            body: 'appended',
          },
          'single line',
        ),
      ).toMatchSnapshot()
    })

    it('handles at_line beyond content length', () => {
      expect(
        injectorV2(
          {
            attributes: {
              at_line: 1000,
            },
            body: 'inserted',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('handles negative at_line', () => {
      expect(
        injectorV2(
          {
            attributes: {
              at_line: -1,
            },
            body: 'inserted',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })
  })

  describe('attribute precedence', () => {
    it('replace takes precedence over all location attributes', () => {
      expect(
        injectorV2(
          {
            attributes: {
              replace: 'rails',
              before: 'nokogiri',
              after: 'source',
              prepend: true,
              append: true,
              at_line: 1,
            },
            body: 'django',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })

    it('skip_if prevents replace operation', () => {
      expect(
        injectorV2(
          {
            attributes: {
              skip_if: 'source',
              replace: 'rails',
            },
            body: 'django',
          },
          gemfile,
        ),
      ).toBe(gemfile) // Should return original content unchanged
    })

    it('at_line takes precedence over other location attributes', () => {
      expect(
        injectorV2(
          {
            attributes: {
              at_line: 3,
              before: 'nokogiri',
              after: 'source',
              prepend: true,
              append: true,
            },
            body: 'inserted at line 3',
          },
          gemfile,
        ),
      ).toMatchSnapshot()
    })
  })
})
