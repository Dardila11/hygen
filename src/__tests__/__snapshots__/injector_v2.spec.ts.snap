// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectorV2 attribute precedence at_line takes precedence over other location attributes 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
inserted at line 3
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 attribute precedence replace takes precedence over all location attributes 1`] = `
"
    source 'http://rubygems.org'
    gem 'django'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 edge cases handles at_line beyond content length 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    
inserted"
`;

exports[`injectorV2 edge cases handles content with only one line 1`] = `
"appended
single line"
`;

exports[`injectorV2 edge cases handles empty body 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'

    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 edge cases handles empty content 1`] = `
"first line
"
`;

exports[`injectorV2 edge cases handles negative at_line 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 eof_last functionality body with newline and eof_last undefined preserves newline 1`] = `
"# Header comment


    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 eof_last functionality body without newline and eof_last true gets newline added 1`] = `
"# Header comment


    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 eof_last functionality if eof_last is false remove empty line from the end of injection body 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 eof_last functionality if eof_last is false with carriage return 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 eof_last functionality if eof_last is true insert empty line to injection body 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen

    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 location-based injection after rails 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 location-based injection append bottom of file 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    gem 'kamikaze' # added by hygen
    "
`;

exports[`injectorV2 location-based injection at_line 2 (below "source") 1`] = `
"
    source 'http://rubygems.org'
    gem 'kamikaze' # added by hygen
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 location-based injection before rails 1`] = `
"
    source 'http://rubygems.org'
    gem 'kamikaze' # added by hygen
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 location-based injection prepend top of file 1`] = `
"    gem 'kamikaze' # added by hygen

    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 multi-line pattern matching correctly interpret multi-line after regex 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 multi-line pattern matching correctly interpret multi-line before regex 1`] = `
"
    source 'http://rubygems.org'
    gem 'kamikaze' # added by hygen
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 multi-line pattern matching handles complex multi-line patterns spanning multiple lines 1`] = `
"
function setup() {
  const config = {
    database: 'mysql',
    host: 'localhost'
  console.log('Setup complete');
  };
  return config;
}
"
`;

exports[`injectorV2 replace functionality replace takes precedence over location attributes 1`] = `
"
config:
  database: mysql
  cache: memcached
  logging: true
"
`;

exports[`injectorV2 replace functionality replaces multiple occurrences with global flag 1`] = `
"
server: postgresql
backup: postgresql
cache: redis
"
`;

exports[`injectorV2 replace functionality replaces single occurrence 1`] = `
"
config:
  database: postgresql
  cache: redis
  logging: true
"
`;

exports[`injectorV2 replace functionality replaces with regex pattern 1`] = `
"{
  \\"name\\": \\"my-app\\",
  \\"version\\": \\"1.0.0\\",
  \\"dependencies\\": {
    \\"react\\": \\"^19.0.0\\"
  }
}"
`;

exports[`injectorV2 skip_if functionality correctly interpret multi-line skip_if regex 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 skip_if functionality skip_if "source" exists 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'nokogiri'
    gem 'httparty'

    "
`;

exports[`injectorV2 skip_if functionality skip_if pattern does not exist - should inject 1`] = `
"
    source 'http://rubygems.org'
    gem 'rails'
    gem 'kamikaze' # added by hygen
    gem 'nokogiri'
    gem 'httparty'

    "
`;
